scripts/validate_docstrings.py

```diff
@@ -490,9 +491,51 @@ def yields(self):
     @property
     def method_source(self):
         try:
-            return inspect.getsource(self.obj)
+            source = inspect.getsource(self.obj)
         except TypeError:
             return ''
+        # Remove common indentation.
+        first_spaces = re.match(' +', source)
+        if first_spaces:
+            first_spaces = first_spaces.group(0)
+            source = re.sub('^' + first_spaces, '', source)
+            source = re.sub(r'\n' + first_spaces, r'\n', source)
+        return source
+
+    @property
+    def method_returns_something(self):
+        '''
+        Check if the docstrings method can return something.
+
+        Bare returns, returns valued None and returns from nested functions are
+        disconsidered.
+
+        Returns
+        -------
+        bool
+            Whether the docstrings method can return something.
+        '''
+
+        def get_returns_not_on_nested_functions(node):
```

Comment:
So noted your previous comment about walk not providing context, but is there still no other option in ast? Does creating a NodeVisitor subclass not preserve context?

While this wouldn't ever go too deep I'm always a little hesitant I think it's typically good to avoid recursive calls which modify objects as they are generally tougher to reason about that non-recursive / mutable code

Reply:
I read the ast doc in search for an alternative and didn't find anything that could help. 

Regarding NodeVisitor, it allows the redefinition of the '.visitor(node)' method, then walks the tree and calls '.visitor(node)' for each node separately. As I understand, it provides no way to influence the walking of the tree itself, so we could prune the nested functions subtrees.

Even so, wouldn't having to create a subclass bring even more complexity to the code?

Reply:
Furthermore, the `get_returns_not_on_nested_functions(node)` doesn't modify the ast structure.